src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -981,249 +683,288 @@ public InputStream retrieveMetadata(String pid, String formatId)
+                // so we only need to delete the pid refs file and related metadata documents
+                Path absPidRefsPath = getHashStoreRefsPath(pid, HashStoreIdTypes.pid.getName());
+                deleteList.add(FileHashStoreUtility.renamePathForDeletion(absPidRefsPath));
+                // Delete items
+                FileHashStoreUtility.deleteListItems(deleteList);
+                deleteMetadata(pid);
+                String warnMsg = "Cid refs file does not exist for pid: " + pid
+                    + ". Deleted orphan pid refs file and metadata.";
+                logFileHashStore.warn(warnMsg);
+
+            } catch (OrphanRefsFilesException orfe) {
+                // `findObject` throws this exception when the pid and cid refs file exists,
+                // but the actual object being referenced by the pid does not exist
+                Path absPidRefsPath = getHashStoreRefsPath(pid, HashStoreIdTypes.pid.getName());
+                String cidRead = new String(Files.readAllBytes(absPidRefsPath));
+                // Since we must access the cid reference file, the `cid` must be synchronized
+                synchronizeReferencedLockedCids(cidRead);
 
-            } else {
-                // Get permanent address of the actual cid
-                String objShardString = FileHashStoreUtility.getHierarchicalPathString(
-                    DIRECTORY_DEPTH, DIRECTORY_WIDTH, cid
-                );
-                Path expectedRealPath = OBJECT_STORE_DIRECTORY.resolve(objShardString);
+                try {
+                    Path absCidRefsPath =
+                        getHashStoreRefsPath(cidRead, HashStoreIdTypes.cid.getName());
+                    updateRefsFile(pid, absCidRefsPath, "remove");
+                    if (Files.size(absCidRefsPath) == 0) {
+                        deleteList.add(FileHashStoreUtility.renamePathForDeletion(absCidRefsPath));
+                    }
+                    deleteList.add(FileHashStoreUtility.renamePathForDeletion(absPidRefsPath));
+                    // Delete items
+                    FileHashStoreUtility.deleteListItems(deleteList);
+                    deleteMetadata(pid);
+                    String warnMsg = "Object with cid: " + cidRead
+                        + " does not exist, but pid and cid reference file found for pid: " + pid
+                        + ". Deleted pid and cid ref files and metadata.";
+                    logFileHashStore.warn(warnMsg);
 
-                // If file exists, delete it.
-                if (Files.exists(expectedRealPath)) {
-                    Files.delete(expectedRealPath);
+                } finally {
+                    // Release lock
+                    releaseReferencedLockedCids(cidRead);
                 }
+            } catch (PidNotFoundInCidRefsFileException pnficrfe) {
+                // `findObject` throws this exception when both the pid and cid refs file exists
```

Comment:
Does this block need to lock the cid? Is there any reason we can put both pid and cid lock in the top `try {` block?

Reply:
This try block does not need the `cid` because this exception is thrown by `findObject` which attempts to locate the `cid`.